src/Compilers/Core/Portable/SourceGeneration/Nodes/InputNode.cs

```diff
@@ -5,61 +5,61 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.Text;
 using System.Threading;
-using Microsoft.CodeAnalysis.PooledObjects;
 
 namespace Microsoft.CodeAnalysis
 {
     /// <summary>
-    /// Input nodes don't actually do anything. They are just placeholders for the value sources
+    /// Input nodes are the 'root' nodes in the graph, and get their values from the inputs of the driver state table
     /// </summary>
-    /// <typeparam name="T"></typeparam>
+    /// <typeparam name="T">The type of the input</typeparam>
     internal sealed class InputNode<T> : IIncrementalGeneratorNode<T>
     {
-        public NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, NodeStateTable<T> previousTable, CancellationToken cancellationToken)
-        {
-            // the input node doesn't change the table. 
-            // instead the driver manipulates the previous table to contain the current state of the node.
-            // we can just return that state, as it's already up to date.
-            return previousTable;
-        }
-
-        // PROTOTYPE(source-generators): how does this work? we definitely want to be able to add custom comparers to the input nodes
-        // I guess its just a 'compare only' node with this as the input?
-        public IIncrementalGeneratorNode<T> WithComparer(IEqualityComparer<T> comparer) => this;
+        private readonly Func<DriverStateTable.Builder, ImmutableArray<T>> _getInput;
+        private readonly IEqualityComparer<T>? _comparer;
 
-        public NodeStateTable<T> CreateInputTable(NodeStateTable<T> previousTable, T item)
+        public InputNode(Func<DriverStateTable.Builder, ImmutableArray<T>> getInput, IEqualityComparer<T>? comparer = null)
         {
-            // PROTOTYPE(source-generators): we should compare the values, not just assume they were added
-            return NodeStateTable<T>.WithSingleItem(item, EntryState.Added);
+            _getInput = getInput;
+            _comparer = comparer;
         }
 
-        public NodeStateTable<T> CreateInputTable(NodeStateTable<T> previousTable, IEnumerable<T> items)
+        public NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, NodeStateTable<T> previousTable, CancellationToken cancellationToken)
         {
+            var inputItems = _getInput(graphState);
+
             // create a mutable hashset of the new items we can check against
-            PooledHashSet<T> itemsSet = PooledHashSet<T>.GetInstance();
-            foreach (var item in items)
+            HashSet<T> itemsSet = new HashSet<T>(_comparer);
+            foreach (var item in inputItems)
             {
                 itemsSet.Add(item);
```

Comment:
Hmm, what happens if `Add` returns `false` here?

Reply:
We'll end up not creating an entry for the duplicate, and thus user code will only see the one instance. We control all the inputs here though, and I didn't think it was valid for any of them to have dupes. I'll add an assert to confirm.

---
In reply to: [632883079](https://github.com/dotnet/roslyn/pull/53383#discussion_r632883079) [](ancestors = 632883079)